src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -759,6 +760,34 @@ PurgeConnectionForPlacement(ShardPlacement *placement)
 
 
 /*
+ * Simply removes a given connection from the transaction participant hash.
+ */
+static void
+RemoveXactConnection(PGconn *connection)
```

Comment:
Hm, reuse of RemoveXactConnection imo makes the error handling behaviour to understand.